File: app/src/main/java/salve/core/ModelZipExtractor.java
package salve.core;

import android.content.Context;
import android.os.StatFs;
import android.util.Log;

import java.io.File;
import java.util.Locale;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import net.lingala.zip4j.ZipFile;
import net.lingala.zip4j.exception.ZipException;
import net.lingala.zip4j.progress.ProgressMonitor;

/**
 * Utilidad para descomprimir modelos ZIP usando Zip4j con soporte Zip64.
 */
public final class ModelZipExtractor {

    private static final String TAG = "ModelZipExtractor";

    private static final int BUF = 256 * 1024;
    private static final long MIN_ZIP_BYTES = 1L << 20;
    private static final long MIN_MARGIN_BYTES = 50L << 20;
    private static final double MARGIN_RATIO = 0.05;

    private static final ExecutorService EXECUTOR = Executors.newSingleThreadExecutor();

    public interface Listener {
        default void onProgress(int percent) {
        }

        void onComplete(File targetDir);

        void onError(Exception e);
    }

    private ModelZipExtractor() {
    }

    public static File extractSync(Context ctx, File zipFile, String prefixLower) throws Exception {
        validateZip(zipFile);

        File targetDir = ensureTargetDir(ctx, prefixLower);
        if (targetHasContent(targetDir)) {
            return targetDir;
        }

        ensureFreeSpace(zipFile, targetDir);

        ModelConsoleOverlay.show();
        ModelConsoleOverlay.log("📦 Descomprimiendo…");

        ZipFile zf = new ZipFile(zipFile);
        zf.setBufferSize(BUF);
        zf.setRunInThread(false);
        try {
            zf.extractAll(targetDir.getAbsolutePath());
            Log.d(TAG, "Modelo descomprimido en: " + targetDir.getAbsolutePath());
            ModelConsoleOverlay.log("✅ Modelo descomprimido: " + prefixLower);
            ModelConsoleOverlay.hideDelayed(1200);
            return targetDir;
        } catch (ZipException e) {
            Log.e(TAG, "Error extrayendo ZIP: " + e.getMessage(), e);
            ModelConsoleOverlay.log("⚠ Error al descomprimir: " + e.getMessage());
            ModelConsoleOverlay.hideDelayed(2500);
            throw e;
        }
    }

    public static void extractAsync(Context ctx, File zipFile, String prefixLower, Listener cb) {
        EXECUTOR.execute(() -> {
            try {
                validateZip(zipFile);

                File targetDir = ensureTargetDir(ctx, prefixLower);
                if (targetHasContent(targetDir)) {
                    if (cb != null) {
                        cb.onComplete(targetDir);
                    }
                    return;
                }

                ensureFreeSpace(zipFile, targetDir);

                ModelConsoleOverlay.show();
                ModelConsoleOverlay.log("📦 Descomprimiendo…");

                ZipFile zf = new ZipFile(zipFile);
                zf.setBufferSize(BUF);
                zf.setRunInThread(true);
                ProgressMonitor pm = zf.getProgressMonitor();
                zf.extractAll(targetDir.getAbsolutePath());

                while (pm.getState() == ProgressMonitor.State.BUSY) {
                    if (cb != null) {
                        cb.onProgress(pm.getPercentDone());
                    }
                    try {
                        Thread.sleep(200);
                    } catch (InterruptedException ignored) {
                        Thread.currentThread().interrupt();
                        break;
                    }
                }

                if (pm.getResult() == ProgressMonitor.Result.ERROR) {
                    throw new RuntimeException("Error extrayendo ZIP", pm.getException());
                }

                if (cb != null) {
                    cb.onComplete(targetDir);
                }
                ModelConsoleOverlay.log("✅ Modelo descomprimido: " + prefixLower);
                ModelConsoleOverlay.hideDelayed(1200);
            } catch (Exception e) {
                if (cb != null) {
                    cb.onError(e);
                }
                ModelConsoleOverlay.log("⚠ Error al descomprimir: " + e.getMessage());
                ModelConsoleOverlay.hideDelayed(2500);
            }
        });
    }

    private static void validateZip(File zipFile) {
        if (zipFile == null || !zipFile.exists()) {
            throw new IllegalArgumentException("ZIP inválido o inexistente");
        }
        if (zipFile.length() < MIN_ZIP_BYTES) {
            throw new IllegalStateException("Archivo ZIP demasiado pequeño: " + zipFile.length());
        }
    }

    private static File ensureTargetDir(Context ctx, String prefixLower) {
        File modelsDir = ModelStore.dir(ctx);
        File targetDir = new File(modelsDir, prefixLower);
        if (!targetDir.exists() && !targetDir.mkdirs()) {
            throw new RuntimeException("No se pudo crear carpeta de modelo: " + targetDir.getAbsolutePath());
        }
        return targetDir;
    }

    private static boolean targetHasContent(File targetDir) {
        if (!targetDir.exists() || !targetDir.isDirectory()) {
            return false;
        }
        File[] existing = targetDir.listFiles();
        return existing != null && existing.length > 0;
    }

    private static void ensureFreeSpace(File zipFile, File targetDir) {
        long zipSize = zipFile.length();
        long estimatedExtractSize = zipSize * 2;
        long margin = Math.max(MIN_MARGIN_BYTES, (long) (estimatedExtractSize * MARGIN_RATIO));
        long required = estimatedExtractSize + margin;
        long free = getAvailableBytes(targetDir);
        if (free < required) {
            throw new RuntimeException(String.format(Locale.US,
                    "Espacio insuficiente: libre=%d bytes, requerido=%d bytes (+%d margen)",
                    free, estimatedExtractSize, margin));
        }
    }

    private static long getAvailableBytes(File path) {
        try {
            StatFs stat = new StatFs(path.getAbsolutePath());
            return stat.getAvailableBlocksLong() * stat.getBlockSizeLong();
        } catch (Throwable t) {
            return Long.MAX_VALUE;
        }
    }
}

File: app/src/main/java/salve/core/ModelZipExtractor.kt
package salve.core

import android.content.Context
import android.util.Log
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.channelFlow
import kotlinx.coroutines.flow.flowOn
import kotlinx.coroutines.withContext
import net.lingala.zip4j.ZipFile
import net.lingala.zip4j.exception.ZipException
import net.lingala.zip4j.progress.ProgressMonitor
import java.io.File
import java.util.Locale

/**
 * Utilidad para descomprimir modelos ZIP usando Zip4j con soporte Zip64.
 */
class ModelZipExtractor(
    private val bufferSize: Int = DEFAULT_BUFFER_SIZE
) {
    suspend fun extract(context: Context, zipFile: File, prefixLower: String): File {
        return withContext(Dispatchers.IO) {
            extractInternal(context, zipFile, prefixLower)
        }
    }

    fun extractWithProgress(context: Context, zipFile: File, prefixLower: String): Flow<ExtractProgress> {
        return channelFlow {
            withContext(Dispatchers.IO) {
                try {
                    validateZip(zipFile)
                    val targetDir = ensureTargetDir(context, prefixLower)
                    if (targetHasContent(targetDir)) {
                        trySend(ExtractProgress.Completed(targetDir))
                        return@withContext
                    }

                    ensureFreeSpace(zipFile, targetDir)

                    ModelConsoleOverlay.show()
                    ModelConsoleOverlay.log("📦 Descomprimiendo…")

                    val zf = ZipFile(zipFile)
                    zf.setBufferSize(bufferSize)
                    zf.setRunInThread(true)
                    val pm = zf.progressMonitor
                    zf.extractAll(targetDir.absolutePath)

                    while (pm.state == ProgressMonitor.State.BUSY) {
                        trySend(ExtractProgress.Progress(pm.percentDone))
                        delay(200)
                    }

                    if (pm.result == ProgressMonitor.Result.ERROR) {
                        throw RuntimeException("Error extrayendo ZIP", pm.exception)
                    }

                    trySend(ExtractProgress.Completed(targetDir))
                    ModelConsoleOverlay.log("✅ Modelo descomprimido: $prefixLower")
                    ModelConsoleOverlay.hideDelayed(1200)
                } catch (e: Exception) {
                    trySend(ExtractProgress.Error(e))
                    ModelConsoleOverlay.log("⚠ Error al descomprimir: ${e.message}")
                    ModelConsoleOverlay.hideDelayed(2500)
                }
            }
        }.flowOn(Dispatchers.IO)
    }

    private fun extractInternal(context: Context, zipFile: File, prefixLower: String): File {
        validateZip(zipFile)

        val targetDir = ensureTargetDir(context, prefixLower)
        if (targetHasContent(targetDir)) {
            return targetDir
        }

        ensureFreeSpace(zipFile, targetDir)

        ModelConsoleOverlay.show()
        ModelConsoleOverlay.log("📦 Descomprimiendo…")

        val zf = ZipFile(zipFile)
        zf.setBufferSize(bufferSize)
        zf.setRunInThread(false)
        return try {
            zf.extractAll(targetDir.absolutePath)
            Log.d(TAG, "Modelo descomprimido en: ${targetDir.absolutePath}")
            ModelConsoleOverlay.log("✅ Modelo descomprimido: $prefixLower")
            ModelConsoleOverlay.hideDelayed(1200)
            targetDir
        } catch (e: ZipException) {
            Log.e(TAG, "Error extrayendo ZIP: ${e.message}", e)
            ModelConsoleOverlay.log("⚠ Error al descomprimir: ${e.message}")
            ModelConsoleOverlay.hideDelayed(2500)
            throw e
        }
    }

    private fun validateZip(zipFile: File?) {
        require(!(zipFile == null || !zipFile.exists())) { "ZIP inválido o inexistente" }
        check(zipFile.length() >= MIN_ZIP_BYTES) { "Archivo ZIP demasiado pequeño: ${zipFile.length()}" }
    }

    private fun ensureTargetDir(context: Context, prefixLower: String): File {
        val modelsDir = ModelStore.dir(context)
        val targetDir = File(modelsDir, prefixLower)
        if (!targetDir.exists() && !targetDir.mkdirs()) {
            throw RuntimeException("No se pudo crear carpeta de modelo: ${targetDir.absolutePath}")
        }
        return targetDir
    }

    private fun targetHasContent(targetDir: File): Boolean {
        if (!targetDir.exists() || !targetDir.isDirectory) {
            return false
        }
        val existing = targetDir.listFiles()
        return existing != null && existing.isNotEmpty()
    }

    private fun ensureFreeSpace(zipFile: File, targetDir: File) {
        val requirement = StorageSpaceUtil.calculateRequirement(zipFile.length(), targetDir)
        if (requirement.availableBytes < requirement.requiredBytes) {
            throw RuntimeException(
                String.format(
                    Locale.US,
                    "Espacio insuficiente: libre=%d bytes, requerido=%d bytes (+%d margen)",
                    requirement.availableBytes,
                    requirement.estimatedExtractBytes,
                    requirement.marginBytes
                )
            )
        }
    }

    sealed class ExtractProgress {
        data class Progress(val percent: Int) : ExtractProgress()
        data class Completed(val targetDir: File) : ExtractProgress()
        data class Error(val error: Exception) : ExtractProgress()
    }

    companion object {
        private const val TAG = "ModelZipExtractor"
        private const val DEFAULT_BUFFER_SIZE = 256 * 1024
        private const val MIN_ZIP_BYTES = 1L shl 20

        @JvmStatic
        @JvmOverloads
        fun extractSync(context: Context, zipFile: File, prefixLower: String, bufferSize: Int = DEFAULT_BUFFER_SIZE): File {
            return kotlinx.coroutines.runBlocking {
                ModelZipExtractor(bufferSize).extract(context, zipFile, prefixLower)
            }
        }
    }
}
